java/arcs/core/data/proto/RecipeProtoDecoder.kt

```diff
@@ -23,6 +23,10 @@ fun RecipeProto.decode(particleSpecs: Map<String, ParticleSpec>): Recipe {
             "Duplicate handle '${it.name}' when decoding recipe '$name'."
         }
     }
+    // Support joins: Associate handles with each other
+    recipeHandles.values.forEach {
```

Comment:
Rather than doing this, try instead:
1) Going over handles that are not `join` handles and decode them first, store them in a map keyed by name.
2) Go again over the list and decode `join` handles, at this point the associated handles should be decoded and present in the map, so pick them up from there.

For now it's ok to assume that we don't compose joins into downstream joins (though that's an interesting idea). If we do that, then we'll need to do this in a loop or do topological sorting, but don't worry for now.

Reply:
Thanks! impl'd two pass system.